src/pyuvdata/uvbeam/beam_interface.py

```diff
@@ -0,0 +1,244 @@
+# Copyright (c) 2022 Radio Astronomy Software Group
+# Licensed under the 2-clause BSD License
+
+"""Definition for BeamInterface object."""
+
+from __future__ import annotations
+
+import copy
+import warnings
+from typing import Any, Literal
+
+import numpy as np
+import numpy.typing as npt
+
+from .analytic_beam import AnalyticBeam
+from .uvbeam import UVBeam
+
+# Other methods we may want to include:
+#  - beam area
+#  - beam squared area
+#  - efield_to_pstokes
+
+
+class BeamInterface:
+    """
+    Definition for a unified beam interface.
+
+    This object provides a unified interface for UVBeam and AnalyticBeam objects
+    to compute beam response values in any direction.
+
+    Attributes
+    ----------
+    beam : pyuvdata.UVBeam or pyuvdata.AnalyticBeam
+        Beam object to use for computations
+    beam_type : str
+        The beam type, either "efield" or "power".
+    include_cross_pols : bool
+        Option to include the cross polarized beams (e.g. xy and yx or en and ne).
+        Used if beam is a UVBeam and and the input UVBeam is an Efield beam but
+        beam_type is "power".
+        Ignored otherwise (the cross pol inclusion is set by the beam object.)
+
+    """
+
+    def __init__(
+        self,
+        beam: AnalyticBeam | UVBeam,
+        beam_type: Literal["efield", "power"] | None = None,
+        include_cross_pols: bool = True,
+    ):
+        if not isinstance(beam, UVBeam) and not issubclass(type(beam), AnalyticBeam):
+            raise ValueError(
+                "beam must be a UVBeam or an AnalyticBeam instance, not a "
+                f"{type(beam)}."
+            )
+        self.beam = beam
+        if isinstance(beam, UVBeam):
+            self._isuvbeam = True
+            if beam_type is None or beam_type == beam.beam_type:
+                self.beam_type = beam.beam_type
+            elif beam_type == "power":
```

Comment:
I think it's ok. The first `if` sets captures the case where we haven't specified a beam_type or we have and it matches the input beam. The `elif` catches the case where we specified `power` but the input is `efield` (has to be `efield` or it would have been caught in the `if` above because they'd match). The `else` captures the case where we specified `efield` but the input is `power`, which isn't allowed.